fish-rust/src/builtins/string.rs

```diff
@@ -0,0 +1,493 @@
+    fn from(_: std::num::TryFromIntError) -> Self {
+        // the result of `try_into` failing on the result of fish_wcsto{i,l}
+        StringError::InvalidValue
+    }
+}
+
+impl From<&'static wstr> for StringError {
+    fn from(s: &'static wstr) -> Self {
+        StringError::InvalidArgs(s)
+    }
+}
+
+impl StringError {
+    fn print_error(
+        &self,
+        args: &[&wstr],
+        parser: &mut parser_t,
+        streams: &mut io_streams_t,
+        optarg: Option<&wstr>,
+        optind: usize,
+        option_name: &wstr,
+    ) {
+        let cmd = args[0];
+        use StringError::*;
+        match self {
+            InvalidArgs(_) | InvalidValue => {
+                let msg = match *self {
+                    InvalidArgs(s) => s.to_owned(),
+                    _ => sprintf!("Invalid %s value", option_name),
+                };
+                let error_msg = wgettext_fmt!("%ls: %ls '%ls'\n", cmd, msg, optarg.unwrap());
+
+                streams.err.append(L!("string "));
+                streams.err.append(error_msg);
+            }
+            NotANumber => {
+                string_error!(streams, BUILTIN_ERR_NOT_NUMBER, cmd, optarg.unwrap());
+            }
+            UnknownOption => {
+                string_unknown_option(parser, streams, cmd, args[optind - 1]);
+            }
+        }
+    }
+
+    fn retval(&self) -> Option<c_int> {
+        STATUS_INVALID_ARGS
+    }
+}
+
+#[derive(Default, PartialEq, Clone, Copy)]
+enum Direction {
+    #[default]
+    Left,
+    Right,
+}
+
+pub(self) fn width_without_escapes(ins: &wstr, start_pos: usize) -> usize {
+    let mut width = 0i32;
+    for c in ins[start_pos..].chars() {
+        let w = fish_wcwidth_visible(c);
+        // We assume that this string is on its own line,
+        // in which case a backslash can't bring us below 0.
+        if w > 0 || width > 0 {
+            width += w;
+        }
+    }
+    // ANSI escape sequences like \e\[31m contain printable characters. Subtract their width
+    // because they are not rendered.
+    let mut pos = start_pos;
+    while let Some(ec_pos) = ins.slice_from(pos).find_char('\x1B') {
+        pos += ec_pos;
+        if let Some(len) = escape_code_length(ins.slice_from(pos)) {
+            let sub = ins.slice_from(pos).slice_to(len);
+            for c in sub.chars() {
+                width -= fish_wcwidth_visible(c);
+            }
+            // Move us forward behind the escape code,
+            // it might include a second escape!
+            // E.g. SGR0 ("reset") is \e\(B\e\[m in xterm.
+            pos += len - 1;
+        } else {
+            pos += 1;
+        }
+    }
+    // we subtracted less than we added
+    debug_assert!(width >= 0, "line has negative width");
+    return width as usize;
+}
+
+pub(self) fn escape_code_length(code: &wstr) -> Option<usize> {
+    use crate::ffi::escape_code_length_ffi;
+    match escape_code_length_ffi(code.as_ptr()).into() {
```

Comment:
This doesn't look right.
escape_code_length_ffi() takes a zero-terminated wchar_t string.
I think we need to go through W0String here.

Reply:
```suggestion
    use crate::wchar_ffi::wstr_to_u32string;

    match escape_code_length_ffi(wstr_to_u32string(code).as_ptr()).into() {
```